### ISSUE TITLE
Can't add unique index to agents.name

### ISSUE THREAD

**Issue by doncadavona** [antonioribeiro/tracker#279]
I am having problems migrating after this commit: [https://github.com/antonioribeiro/tracker/commit/1e8b7043961b8b36c017a8443ba72472215e6e13](https://github.com/antonioribeiro/tracker/commit/1e8b7043961b8b36c017a8443ba72472215e6e13):

```sql
SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'name' used in key specification without a key length (SQL: alter table `tracker_agents` add unique `tracker_agents_name_unique`(`name`))
```

As explained in [http://stackoverflow.com/a/1827099/3936053](http://stackoverflow.com/a/1827099/3936053), MySQL can't add unique indexes to texts.

As suggestion, we could add a `hash` column of type `varchar(64)` or `string(64)` for the hash of each user agents and make that hash column as the one unique, instead of the `name` column of type `text`. In this case, we can store long user agent texts without worrying about properly indexing their uniqueness as discussed in [http://stackoverflow.com/a/14033466/3936053](http://stackoverflow.com/a/14033466/3936053).

Update: June 2, 2017

Found this approach was also suggested here: [http://stackoverflow.com/a/14033466/3936053](https://stackoverflow.com/a/13210391/3936053)

**Comment by Antipov007** [comment]
I have same error when try doing migration. I tried `php artisan migrate --database=tracker`  and `php artisan migrate ` and in both cases i have this error:
"SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'name' used in key specification without a key length (SQL: alter table `tracker_agents` add unique `tracker_agents_name_unique`(`name`))"
MYSQL 5.7
DB Settings:
        'tracker' => [
            'driver' => 'mysql',
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => env('DB_DATABASE', 'forge'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'unix_socket' => env('DB_SOCKET', ''),
            'charset' => 'utf8mb4',
            'collation' => 'utf8mb4_unicode_ci',
            'prefix' => '',
            'strict' => true,
            'engine' => null,
            ],
It create almost all tables but can not complete because get this error((
Maybe somebody deside this problem?

**Comment by Antipov007** [comment]
WOW FOUND DECISSION!!!! THIS IS REPOST:

> **nipun5perera** commented on **25 Feb**
> 
> I did edit create_tracker_agents_table.php and changed ,
> 
> $table->string('name')->unique();
> 
> to
> 
> $table->string('name',255);
> 
> and then added
> 
> $table->unique('name');
> 
> so my create would look like this,
> 
> $table->bigIncrements('id');
> $table->string('name',255);
> $table->string('browser')->index();
> $table->string('browser_version');
> $table->timestamp('created_at')->index();
> $table->timestamp('updated_at')->index();
> $table->unique('name');
> 
> and then edited fix_agent_name.php and changed
> 
> $table->mediumText('name')->unique()->change();
> 
> to
> 
> $table->string('name',255)->unique()->change();
> 
> for me this solved that key length problem and everything migrated successfully.

**Comment by doncadavona** [comment]
That works fine @Antipov007 , although your app may still encounter user agents that exceed your fields' length (255) along the way. But depending on the users of your app, you may or may not encounter that.
I see the best fix would be to implement the hashing of user agents. I'll see if I can submit a PR for that soon. Until then, your fix definitely works.

**Comment by antonioribeiro** [comment]
Should be solved now, on 3.1.4.

**Comment by doncadavona** [comment]
Thanks @antonioribeiro !
